### PR TITLE
[SYCL][L0][PI] Add support for link copy engines

### PR DESCRIPTION
**Pull request by asudarsa**
Signed-off-by: Arvind Sudarsanam <arvind.sudarsanam@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 1]
```diff
@@ -584,6 +615,11 @@ struct _pi_queue : _pi_object {
     bool InUse;
     // Record if the associated command list (if any) is a "copy" command list.
     bool IsCopyCommandList;
+    // Record the index of copy queue (in the vector of available copy queues)
+    // to which the command list (if any) will be submitted.
+    // If there is no command list, or if the command list is not a copy command
+    // list, the value is set to -1.
+    int CopyQueueIndex;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:622]: can't you use CopyQueueIndex == -1 check to replace the IsCopyCommandList member?

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 2]
```diff
@@ -496,21 +507,41 @@ const pi_uint32 DynamicBatchStartSize = 4;
...
+  // One of the many available copy command queues will be used for
+  // submitting command lists to. This variable stores index of the last used
+  // copy command queue in the ZeCopyCommandQueues vector.
+  int32_t LastUsedCopyCommandQueueIndex = -1;
+
+  // This function will return one of possibly multiple available copy queues.
+  // Currently, a round robin strategy is used.
+  // It will return nullptr if no copy command queues are available for use.
+  ze_command_queue_handle_t
+  getZeCopyCommandQueue(int &CopyQueueIndex,
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:538]: I suggest to make first argument a pointer for consistency

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -945,13 +976,16 @@ pi_result _pi_queue::executeCommandList(ze_command_list_handle_t ZeCommandList,
...
-
-  bool UseCopyEngine = getZeCommandListIsCopyList(ZeCommandList);
+  int Index = -1;
+  bool UseCopyEngine = getZeCommandListIsCopyList(ZeCommandList, Index);
   if (UseCopyEngine)
     zePrint("Command list to be executed on copy engine\n");
+  // If available, get the copy command queue assosciated with
+  // ZeCommandList
+  auto ZeCopyCommandQueue =
+      (Index == -1) ? nullptr : ZeCopyCommandQueues[Index];
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:986]: Can't you replace getZeCommandListIsCopyList with getCopyQueueIndex and remove the earlier?

**Review by smaslov-intel — COMMENTED** [review]
Please address comments

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 2 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp]: Addressed this in the next patch. Thanks

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Done. Good optimization.

File: sycl/plugins/level_zero/pi_level_zero.hpp (hunk 1 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.hpp:615]: Yes. I think we can.

**Review by smaslov-intel — APPROVED** [review]
LGTM

**Comment by romanovvlad** [comment]
Can we start writing component tests for L0 plugin? It could be applications that call PI API of plugins and checks that calls to L0 API are correct(similar to how we check that SYCL API is lowered correctly to PI API).

**Comment by asudarsa** [comment]
Hi @romanovvlad 
I think I can open a separate Jira for this and track it from there. Is that acceptable?

Thanks

**Comment by romanovvlad** [comment]
> 
> 
> Hi @romanovvlad
> I think I can open a separate Jira for this and track it from there. Is that acceptable?
> 
> Thanks

Ok.

**Pull request merged**